File: openreview/venue/process/invite_assignment_pre_process.js
async function process(client, edge, invitation) {
  client.throwErrors = true

  const { groups } = await client.getGroups({ id: edge.domain })
  const domain = groups[0]
  const reviewersId = invitation.content.match_group?.value
  const assignmentInvitationId = invitation.content.assignment_invitation_id?.value
  const conflictInvitationId = invitation.content.conflict_invitation_id?.value
  const assignmentLabel = invitation.content.assignment_label?.value
  const committeeRole = invitation.content.committee_role?.value
  const inviteLabel = invitation.content.invite_label?.value
  const committeeName = reviewersId?.split("/")?.pop()
  const conflictPolicy = domain.content?.[`${committeeRole}_conflict_policy`]?.value
  const conflictNYears = domain.content?.[`${committeeRole}_conflict_n_years`]?.value
  const quota = domain.content?.[`submission_assignment_max_${committeeRole}`]?.value

  if (edge.ddate && edge.label !== inviteLabel) {
    return Promise.reject(new OpenReviewError({ name: 'Error', message: `Cannot cancel the invitation since it has status: "${edge.label}"` }))
  }

  if (edge.label !== inviteLabel) {
    return
  }

  const { notes } = await client.getNotes({ id: edge.head })
  const submission = notes[0]

  const profiles = await client.tools.getProfiles([edge.tail], true)
  const userProfile = profiles[0]

  if (userProfile.id !== edge.tail) {
    const { edges } = await client.getEdges({ invitation: edge.invitation, head: edge.head, tail: userProfile.id })
    if (edges.length) {
      return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite ${userProfile.id}, the user is already invited` }))
    }
  }

  const { edges } = await client.getEdges({ invitation: assignmentInvitationId, head: edge.head, tail: userProfile.id, label: assignmentLabel })
  if (edges.length) {
    return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite ${userProfile.id}, the user is already assigned` }))
  }

  if (quota) {
    const acceptLabel = invitation?.content?.accepted_label?.value ?? '';
    const invitedLabel = invitation?.content?.invited_label?.value ?? '';
    const pendingSignUpLabel = 'Pending Sign Up';
    const includedLabels = [acceptLabel, invitedLabel, pendingSignUpLabel];
    
    const [{ edges: inviteAssignmentEdges }, { edges: assignmentEdges }] = await Promise.all([
      client.getEdges({ invitation: edge.invitation, head: edge.head }),
      client.getEdges({ invitation: assignmentInvitationId, head: edge.head })
    ])

    // Convert includedLabels to lowercase for case-insensitive comparison
    const lowerCaseIncludedLabels = includedLabels.map(label => label.toLowerCase());
    const assignmentTails = assignmentEdges.map(e => e.tail);
    
    // Filter invite assignment edges to include only edges that contain any of the includedLabels as substrings (case-insensitive) and exclude the current edge.id
    const filteredInviteAssignmentEdges = inviteAssignmentEdges.filter(e => {
      const edgeLabel = e?.label?.toLowerCase() ?? '';
      // Check if edgeLabel includes any of the includedLabels
      const includesIncludedLabel = lowerCaseIncludedLabels.some(includedLabel => edgeLabel.includes(includedLabel));
      // Exclude if it already has a corresponding assignment edge (same tail)
      const hasCorrespondingAssignment = assignmentTails.includes(e.tail);
      // Include edge only if it contains any of the includedLabels, is not the current edge, and doesn't have a corresponding assignment
      return includesIncludedLabel && e.id !== edge.id && !hasCorrespondingAssignment;
    });
    const alreadyPosted = inviteAssignmentEdges.filter(e => e.id === edge.id).length > 0

    if (!alreadyPosted && filteredInviteAssignmentEdges.length + assignmentEdges.length >= quota) {
      return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite assignment, total assignments and invitations must not exceed ${quota}; invite edges=${filteredInviteAssignmentEdges.length} assignment edges=${assignmentEdges.length}` }))
    }
  }

  if (userProfile.id.startsWith('~')) {
    const { groups: reviewersGroup } = await client.getGroups({ id: reviewersId, member: userProfile.id })

    if (reviewersGroup.length) {
      const { edges } = await client.getEdges({ invitation: conflictInvitationId, head: edge.head, tail: userProfile.id })
      if (edges.length) {
        return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite ${userProfile.id}, the user has a conflict` }))
      }

      if (assignmentLabel) {
        return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite ${userProfile.id}, the user is an official reviewer` }))
      }
    }
  }

  const authorProfiles = await client.tools.getProfiles(submission.content.authorids?.value, true)
  const conflicts = await client.tools.getConflicts(authorProfiles, userProfile, conflictPolicy, conflictNYears)
  if (conflicts.length) { 
    return Promise.reject(new OpenReviewError({ name: 'Error', message: `Can not invite ${userProfile.id}, the user has a conflict` }))
  }
}

File: openreview/venue/process/invite_assignment_pre_process.py
def process(client, edge, invitation):

    domain = client.get_group(edge.domain)
    reviewers_id = invitation.content['match_group']['value']
    assignment_invitation_id = invitation.content['assignment_invitation_id']['value']
    assignment_label = invitation.content.get('assignment_label', {}).get('value')
    invite_label = invitation.content['invite_label']['value']
    reviewers_conflict_id = domain.content.get('reviewers_conflict_id', {}).get('value')
    conflict_policy = domain.content.get('reviewers_conflict_policy', {}).get('value', 'Default')
    conflict_n_years = domain.content.get('reviewers_conflict_n_years', {}).get('value')
    print(edge.id)

    if edge.ddate is None and edge.label == invite_label:

        ## Get the submission
        notes=client.get_notes(id=edge.head)
        if not notes:
            raise openreview.OpenReviewException(f'Note not found: {edge.head}')
        submission=notes[0]

        ## - Get profile
        user = edge.tail
        print(f'Get profile for {user}')
        user_profile=openreview.tools.get_profiles(client, [user], with_publications=True, with_relations=True)[0]

        if user_profile:
            if user_profile.id != user:
                ## - Check if the user is already invited
                edges=client.get_edges(invitation=edge.invitation, head=edge.head, tail=user_profile.id)
                if edges:
                    raise openreview.OpenReviewException(f'Already invited as {edges[0].tail}')

            ## - Check if the user is already assigned
            edges=client.get_edges(invitation=assignment_invitation_id, head=edge.head, tail=user_profile.id, label=assignment_label)
            if edges:
                raise openreview.OpenReviewException(f'Already assigned as {edges[0].tail}')

            ## - Check if the user is an official reviewer
            if user_profile.id.startswith('~') and client.get_groups(id=reviewers_id, member=user_profile.id):

                ## - Check if the user has a conflict
                edges=client.get_edges(invitation=reviewers_conflict_id, head=edge.head, tail=user_profile.id)
                if edges:
                    raise openreview.OpenReviewException(f'Conflict detected for {user_profile.get_preferred_name(pretty=True)}')

                ## Only check this when there are proposed assignments
                if assignment_label:
                    raise openreview.OpenReviewException(f'Reviewer {user_profile.get_preferred_name(pretty=True)} is an official reviewer, please use the "Assign" button to make the assignment.')


        else:
            if user.startswith('~'):
                raise openreview.OpenReviewException(f'Profile not found {user}')
            user_profile=openreview.Profile(id=user,
                content={
                    'names': [],
                    'emails': [user],
                    'preferredEmail': user
                })


        print(f'Check conflicts for {user_profile.id}')
        ## - Check conflicts
        authorids = submission.content['authorids']['value']
        author_profiles = openreview.tools.get_profiles(client, authorids, with_publications=True, with_relations=True)
        conflicts=openreview.tools.get_conflicts(author_profiles, user_profile, policy=conflict_policy, n_years=conflict_n_years)
        if conflicts:
            print('Conflicts detected', conflicts)
            raise openreview.OpenReviewException(f'Conflict detected for {user_profile.get_preferred_name(pretty=True)}')

    return edge